src/Catalog/Persistence/AzureStorage.cs

```diff
@@ -379,21 +380,57 @@ public override Uri GetUri(string name)
             var source = new CloudBlockBlob(firstResourceUri);
+            var destination = GetBlockBlobReference(GetName(secondResourceUri));
 
             // For interacting with the source, we just use the same blob request options as the destination blob.
             ApplyBlobRequestOptions(source);
 
-            if (await destination.ExistsAsync())
+            return await AreSynchronized(new AzureCloudBlockBlob(source), new AzureCloudBlockBlob(destination));
+        }
+
+        public async Task<bool> AreSynchronized(ICloudBlockBlob sourceBlockBlob, ICloudBlockBlob destinationBlockBlob)
+        {
+            if (await destinationBlockBlob.ExistsAsync(CancellationToken.None))
             {
-                if (await source.ExistsAsync())
+                if (await sourceBlockBlob.ExistsAsync(CancellationToken.None))
                 {
-                    return !string.IsNullOrEmpty(source.Properties.ContentMD5) && source.Properties.ContentMD5 == destination.Properties.ContentMD5;
+                    var sourceBlobMetadata = await sourceBlockBlob.GetMetadataAsync(CancellationToken.None);
+                    var destinationBlobMetadata = await destinationBlockBlob.GetMetadataAsync(CancellationToken.None);
+                    if (sourceBlobMetadata == null || destinationBlobMetadata == null)
+                    {
+                        return false;
+                    }
+
+                    var sourceBlobHasSha512Hash = sourceBlobMetadata.TryGetValue(Sha512HashAlgorithmId, out var sourceBlobSha512Hash);
+                    var destinationBlobHasSha512Hash = destinationBlobMetadata.TryGetValue(Sha512HashAlgorithmId, out var destinationBlobSha512Hash);
+                    if (!sourceBlobHasSha512Hash)
+                    {
+                        Trace.TraceWarning(string.Format("The source blob ({0}) doesn't have the SHA512 hash.", sourceBlockBlob.Uri.ToString()));
+                    }
+                    if (!destinationBlobHasSha512Hash)
+                    {
+                        Trace.TraceWarning(string.Format("The destination blob ({0}) doesn't have the SHA512 hash.", destinationBlockBlob.Uri.ToString()));
+                    }
+                    if (sourceBlobHasSha512Hash && destinationBlobHasSha512Hash)
+                    {
+                        if (sourceBlobSha512Hash == destinationBlobSha512Hash)
+                        {
+                            Trace.WriteLine(string.Format("The source blob ({0}) and destination blob ({1}) have the same SHA512 hash and are synchronized.",
+                                sourceBlockBlob.Uri.ToString(), destinationBlockBlob.Uri.ToString()));
+                            return true;
+                        }
+
+                        Trace.TraceWarning(string.Format("The source blob ({0}) and destination blob ({1}) have the different SHA512 hash and are not synchronized. " +
```

Comment:
>TraceWarning [](start = 30, length = 12)

Should this be a `TraceWarning`? Blob mutations seem like a normal operation, considering making this a `WriteLine`

Reply:
Yeah, because it's an unexpected behavior, so we make it as a warning.

Reply:
Ah okay. Maybe add a comment since that's not evident